master/internal/rm/agentrm/agent.go

```diff
@@ -103,373 +110,448 @@ type (
+	}
+	a.agentState.deallocateContainer(msg.containerID)
+	return nil
+}
 
-		if a.awaitingRestore {
-			a.awaitingRestore = false
-		}
+func (a *agent) State() (*agentState, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 
-		a.socket.Outbox <- masterSetAgentOptions
+	if !a.started {
+		return nil, errors.New("agent state is not available: agent not started")
+	}
 
-		if a.awaitingReconnect {
-			ctx.Log().Info("agent reconnected")
-			a.awaitingReconnect = false
+	return a.agentState.deepCopy(), nil
+}
 
-			// Cancel reconnect timers.
-			for _, timerActor := range a.reconnectTimers {
-				timerActor.Stop()
-			}
-			a.reconnectTimers = nil
+func (a *agent) StartTaskContainer(msg sproto.StartTaskContainer) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 
-			// Re-propagate our old state back on successful recovery.
-			if a.preDisconnectEnabled {
-				a.agentState.enable()
-			} else {
-				a.agentState.disable(a.preDisconnectDraining)
-			}
-			a.agentState.patchAllSlotsState(patchAllSlotsState{
-				enabled: &a.agentState.enabled,
-				drain:   &a.agentState.draining,
-			})
+	a.startTaskContainer(msg)
+}
 
-			if len(a.reconnectBacklog) > 0 {
-				for i, msg := range a.reconnectBacklog {
-					ctx.Log().Debugf("will replay reconnectBacklog %d %s", i, reflect.TypeOf(msg))
-				}
-			}
+func (a *agent) startTaskContainer(msg sproto.StartTaskContainer) {
+	if a.awaitingReconnect {
+		a.bufferForRecovery(msg)
+		return
+	}
+	log := a.syslog.
+		WithFields(msg.LogContext.Fields()).
+		WithField("container-id", msg.StartContainer.Container.ID).
+		WithField("slots", len(msg.StartContainer.Container.Devices))
+	log.Infof("starting container")
 
-			for _, msg := range a.reconnectBacklog {
-				if err := a.receive(ctx, msg); err != nil {
-					ctx.Log().WithError(err).WithField("msg", msg).Errorf("replaying backlog")
-					return errors.Wrapf(err, "replaying backlog")
-				}
-			}
-			a.reconnectBacklog = nil
-			ctx.Tell(a.resourcePool, sproto.UpdateAgent{Agent: ctx.Self()})
-		}
+	a.socket.Outbox <- aproto.AgentMessage{StartContainer: &msg.StartContainer}
 
-	case socketDisconnect:
-		defer ctx.Tell(a.resourcePool, sproto.UpdateAgent{Agent: ctx.Self()})
-		defer a.socketDisconnected(ctx)
+	if err := a.agentState.startContainer(msg); err != nil {
+		log.WithError(err).Error("failed to update agent state")
+	}
+}
 
-		err := a.socket.Close()
-		if err != nil {
-			if !a.started {
-				// If we happen to fail before the agent has started and been registered with
-				// the resource manager, then nothing can be running on it. In this case we
-				// just fail outright and make it restart.
-				return errors.Wrapf(err, "child failed: %s", a.socket.Name())
-			}
+func (a *agent) KillTaskContainer(msg sproto.KillTaskContainer) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 
-			ctx.Log().
-				WithError(err).
-				Errorf("WebSocket failed, awaiting reconnect: %s", a.socket.Name())
-			return nil
-		}
+	a.killTaskContainer(msg)
```

Comment:
question: why are you wrapping a lock around the `killTaskContainer` helper function instead of inserting that code block directly here?

Reply:
I took a really methodical path refactoring this, minimizing changes. I'll come back and do a style-only refactor to undo some stuff like this (I don't think this one is particularly offensive, though).